### PR TITLE
Refactor/remove plain groth16 prover

### PR DESCRIPTION
**Pull request by 0xThemis**
This PR has some clean up.
Most notably it adds a verifier feature for co-Groth16. This means the ark::groth16 dep is only used when having this feature enabled.

Additionally, some refactoring to mirror PLONK. We now only have one single prove call to groth16 (we removed the matrices call). We now only take the ZKey and the witness and read the matrices from there, so we never recompute them.

QUESTION: I do not know if we need the R1CS file. We only use it for the witness extension to read the public inputs. We can also do that with the zkey, but snarkjs needs the r1cs file for witness extension therefore I left it. But we could just remove it....

### CODE REVIEW

File: circom-types/src/groth16/verification_key.rs [hunk 1]
```diff
@@ -198,6 +194,7 @@ where
         PreparedVerifyingKey::<P>::from(self)
     }
 }
+    */
```
Comment (dkales) [inline circom-types/src/groth16/verification_key.rs:197]: why are we keeping this comment

File: circom-types/src/groth16/verification_key.rs [hunk 2]
```diff
@@ -291,53 +288,53 @@ mod test {
             fs::read_to_string("../test_vectors/Groth16/bls12_381/verification_key.json").unwrap();
         let vk = serde_json::from_str::<JsonVerificationKey<Bls12_381>>(&vk_string).unwrap();
         let alpha_1 = test_utils::to_g1_bls12_381!(
-            "1006274644424409217186953213662503172434575368717179668374437354164299156533023899663544377042948523039106817870257",
-            "1056550161284843386841851874013894300183177686460359778532448293436751746312447247913808276245741832724930728476526"
+            "573513743870798705896078935465463988747193691665514373553428213826028808426481266659437596949247877550493216010640",
```
Comment (dkales) [inline circom-types/src/groth16/verification_key.rs:291]: why do we need new KAT vectors?

File: circom-types/src/groth16/verification_key.rs [hunk 3]
```diff
@@ -30,32 +26,32 @@ where
...
-    delta_2: P::G2Affine,
+    pub delta_2: P::G2Affine,
     /// The pairing of α and β of the verification key ∈ Gt
     #[serde(rename = "vk_alphabeta_12")]
     #[serde(serialize_with = "P::serialize_gt::<_>")]
     #[serde(deserialize_with = "P::deserialize_gt_element::<_>")]
-    alpha_beta_gt: P::TargetField,
-    /// Referred to as `gamma_abc_g1` in [`ark_groth16::data_structures::VerifyingKey`] and is used to bind the public inputs to the proof
+    pub alpha_beta_gt: P::TargetField,
+    /// Bind the public inputs to the proof
```
Comment (dkales) [inline circom-types/src/groth16/verification_key.rs:50]: Used to bind ...

File: circom-types/src/groth16/zkey.rs [hunk 4]
```diff
@@ -195,98 +215,74 @@ where
...
-
 impl<P: Pairing + CircomArkworksPairingBridge> HeaderGroth<P>
 where
     P::BaseField: CircomArkworksPrimeFieldBridge,
     P::ScalarField: CircomArkworksPrimeFieldBridge,
 {
     fn read<R: Read>(mut reader: &mut R) -> ZKeyParserResult<Self> {
-        // TODO: Impl From<u32> in Arkworks
-        let n8q: u32 = u32::deserialize_uncompressed(&mut reader)?;
+        let _n8q: u32 = u32::deserialize_uncompressed(&mut reader)?;
```
Comment (dkales) [inline circom-types/src/groth16/zkey.rs:251]: Should we assert/err on the expected value of these?

File: collaborative-plonk/src/round1.rs [hunk 5]
```diff
@@ -295,12 +295,13 @@ pub mod tests {
     use std::{fs::File, io::BufReader};
 
     use ark_bn254::Bn254;
-    use circom_types::{groth16::witness::Witness, plonk::ZKey, r1cs::R1CS};
-    use collaborative_groth16::{circuit::Circuit, groth16::SharedWitness};
+    use circom_types::plonk::ZKey;
+    use collaborative_groth16::groth16::SharedWitness;
```
Comment (dkales) [inline collaborative-plonk/src/round1.rs:299]: did we not want to put the sharedwitness into a separate crate? Other PR, right?

File: collaborative-plonk/src/round3.rs [hunk 6]
```diff
@@ -517,7 +517,8 @@ pub mod tests {
     use std::{fs::File, io::BufReader};
 
     use ark_bn254::Bn254;
-    use circom_types::{groth16::witness::Witness, plonk::ZKey};
+    use circom_types::groth16::Witness;
```
Comment (dkales) [inline collaborative-plonk/src/round3.rs:520]: This witness type is also no longer `groth16` only, it  should live 1 lvl higher

File: circom-types/src/groth16/verification_key.rs (hunk 1 above)
Comment (0xThemis) [reply circom-types/src/groth16/verification_key.rs:197]: removed the comment.

File: circom-types/src/groth16/verification_key.rs (hunk 3 above)
Comment (0xThemis) [reply circom-types/src/groth16/verification_key.rs:50]: done

File: circom-types/src/groth16/zkey.rs (hunk 4 above)
Comment (0xThemis) [reply circom-types/src/groth16/zkey.rs:251]: I thought about that but I if everything else is correct it doesnt matter what those values are. Therefore I just ignored them. Maybe put a warning log in there?
We maybe want logging anyways?

File: collaborative-plonk/src/round1.rs (hunk 5 above)
Comment (0xThemis) [reply collaborative-plonk/src/round1.rs:299]: Yes we did wanted to do that, but the plan was to do that in another PR. Also I didn't want to decide alone how we structure everything

File: collaborative-plonk/src/round3.rs (hunk 6 above)
Comment (0xThemis) [reply collaborative-plonk/src/round3.rs:520]: moved it

File: circom-types/src/groth16/verification_key.rs (hunk 2 above)
Comment (0xThemis) [reply circom-types/src/groth16/verification_key.rs:291]: One test we had did not use the .json verification key to verify, but the verification key in the zkey.
I wanted to mirror the actual workflow, therefore tried to read the verification key, but the verification key in the test folder was no longer the one needed for the verification. Therefore, I extracted it from the zkey, overriding the other test case.

I could just keep both, if prefered

Comment (dkales) [reply circom-types/src/groth16/verification_key.rs:236]: Nah was just a question, is fine as is

File: circom-types/src/groth16/zkey.rs (hunk 4 above)
Comment (dkales) [reply circom-types/src/groth16/zkey.rs:251]: We should probably add a lot more tracing in general in a new pr

Comment (dkales) [reply circom-types/src/groth16/zkey.rs:251]: They still have some expected value though and if they are not that then there is some fuckery going on and we should probably error

**Review by dkales — APPROVED** [review]

**Pull request merged**